src/operator/numpy/linalg/np_norm-inl.h

```diff
@@ -0,0 +1,830 @@
+  double ord;
+  dmlc::optional<mxnet::TShape> axis;
+  bool keepdims;
+  int flag;
+  DMLC_DECLARE_PARAMETER(NumpyNormParam) {
+    DMLC_DECLARE_FIELD(ord).set_default(2)
+    .describe("Order of the norm. inf means numpy’s inf object.");
+    DMLC_DECLARE_FIELD(axis).set_default(dmlc::optional<mxnet::TShape>())
+    .describe(R"code(If axis is an integer, it specifies the axis of x along
+     which to compute the vector norms. If axis is a 2-tuple,
+     it specifies the axes that hold 2-D matrices, and the matrix norms of
+     these matrices are computed. If axis is None then either a vector norm (when x is 1-D)
+     or a matrix norm (when x is 2-D) is returned.If axis is an integer,
+     it specifies the axis of x along which to compute the vector norms.
+     If axis is a 2-tuple, it specifies the axes that hold 2-D matrices,
+     and the matrix norms of these matrices are computed. If axis is None then either a
+     vector norm (when x is 1-D) or a matrix norm (when x is 2-D) is returned.)code");
+    DMLC_DECLARE_FIELD(keepdims).set_default(false)
+    .describe("If this is set to `True`, the reduced axis is left "
+    "in the result as dimension with size one.");
+    DMLC_DECLARE_FIELD(flag).set_default(-1)
+    .describe("Mapping relations between ord and flag."
+    "ord:  None,  'fro', 'nuc', 'inf'  '-inf'."
+    "flag:  0 ,    1,      2,    3,      4. ");
+  }
+};
+
+TShape swapMatDims(const TShape &shape, const TShape &axis);
+TShape inverseTranspose(const TShape &axes);
+
+bool NumpyNormShape(const nnvm::NodeAttrs& attrs,
+                    mxnet::ShapeVector *in_attrs,
+                    mxnet::ShapeVector *out_attrs);
+
+template<typename xpu>
+void NumpyLpNormCompute(const nnvm::NodeAttrs& attrs,
+                        const OpContext& ctx,
+                        const std::vector<TBlob>& inputs,
+                        const std::vector<OpReqType>& req,
+                        const std::vector<TBlob>& outputs) {
+  const NumpyNormParam& param = nnvm::get<NumpyNormParam>(attrs.parsed);
+  double ord = param.ord;
+
+  if (req[0] == kNullOp) return;
+
+  mxnet::TShape small;
+  mxnet::TShape out_shape = outputs[0].shape_;
+  if (param.keepdims) {
+    small = outputs[0].shape_;
+  } else {
+    small = ReduceAxesShapeImpl(inputs[0].shape_, param.axis, true, false);
+    const_cast<std::vector<TBlob>&>(outputs)[0] = outputs[0].reshape(small);
+  }
+  bool safe_acc = dmlc::GetEnv("MXNET_SAFE_ACCUMULATION", false);
+  if (!safe_acc && inputs[0].type_flag_ == mshadow::kFloat16) {
+    common::LogOnce("MXNET_SAFE_ACCUMULATION=1 is recommended for LpNorm with float16 inputs. "
+                    "See https://mxnet.apache.org/api/faq/env_var "
+                    "for more details.");
+  }
+  if (param.axis.value().ndim() != 2) {  // elementwise Lp-norm
+    if (ord == -std::numeric_limits<double>::infinity()) {  // -inf norm
+      LOG(FATAL) << "Under construction. ";
```

Comment:
Use a better error message such as "ord = inf is not supported yet.", similar for the case below.

Reply:
The case is actually handled in front-end, these are placeholders in case people want me to move the implementation to the back-end, and would not be reached. I will change the message to "Case handled in front-end" instead.